`lazy v`: optimize structured constants

gasche opened the pull request:
Sometimes we want to build values at a lazy type, but we are not interested in delaying evaluation. A typical example is the function `rev_append_list li s`, which adds the reverse of a strict list `li` to the beginning of a lazy list `s`.

```ocaml
type 'a stream = 'a cell Lazy.t
and 'a cell = Nil | Cons of 'a * 'a stream

let rec rev_append_list li s =
  match li with
  | [] -> s
  | x :: xs -> rev_append_list xs (lazy (Cons (x, s)))
```

The representation of lazy values in OCaml would let us optimize the expression `lazy (Cons (x, s))` by computing `Cons (x, s)` right now, and returning it directly (without using a forward block; there is no risk of confusion here). The compiler optimizes related expressions, such as `lazy Nil`, but it currently does not optimize `lazy (Cons (x, s))`, because we didn't bother detecting structured constants.

The present PR changes the optimization of lazy values to also support structured constants. The program above gets optimized at compile-time to not include any laziness-related indirection.

### Lazy.from_val

The `Lazy` module offers a `Lazy.from_val` helper, which performs a dynamic check to avoid the Forward block when possible. `Lazy.from_val` can be used manually by experts to avoid indirections, for example for structured constants, when the compiler optimizations are not good enough. For example the [BatLazyList module](https://github.com/ocaml-batteries-team/batteries-included/blob/7d2c4f20b28f61450a129aeb886aa28ad5dd7ee7/src/batLazyList.ml#L284) of `batteries` uses it to implement `rev_append_list`. 

The compiler checks for the possibility to avoid the Forward block statically, so it can generate better code than `Lazy.from_val` (for flat values in trunk, and for flat and structured values in this PR).

In my eyes the value of the change is not to be faster than `Lazy.from_val`, but rather to be at least as fast in many cases without requiring users to know about this manual optimization. (and as compact in terms of memory representation, without having to reason about GC-time shortcutting)

Note that not all use-cases of `Lazy.from_val e` are covered here, as users could decide to use it when `e` is not a value and performs effect (then it is equivalent to `let x = e in lazy x`, with dynamic shortcutting of forward blocks), or when `e` is a constant but its type is not statically known. (One could refine the code generator to insert a call to `Lazy.from_val` in cases where the static check fails because the types are unknown at compile-time, but the compiler typically does not know about the standard library.)

### Obligatory micro-benchmark

```ocaml
let () =
  let input = List.init 10_000 Fun.id in
  for i = 0 to 10_000 do
    ignore (Sys.opaque_identity (rev_append_list input (lazy Nil)))
  done
```

I measured 4 versions of this benchmark, along two axes:
- version: 5.3.0 or the current branch
- `lazy` or `Lazy.from_val`: the latter uses a variant of the function above with `Lazy.from_val (Cons (x, s))`

```
  ./test.branch.opt ran
    1.46 ± 0.02 times faster than ./test_from_val.branch.opt
    1.47 ± 0.02 times faster than ./test_from_val.530.opt
    2.67 ± 0.03 times faster than ./test.530.opt
```

The results suggest that `lazy (Cons (x, xs))` with this PR is 2.7x faster than without this PR, and 1.5x faster than `Lazy.from_val`.

gasche committed:
(lazy e): optimize structured constants as Lazy.from_val would do

kit-ty-kate commented on stdlib/option.mli:
```diff
@@ -54,6 +54,14 @@ val fold : none:'a -> some:('b -> 'a) -> 'b option -> 'a
 val iter : ('a -> unit) -> 'a option -> unit
 (** [iter f o] is [f v] if [o] is [Some v] and [()] otherwise. *)
 
+val for_all : ('a -> bool) -> 'a option -> bool
+(** [for_all p None] is [true],
+    [for_all p (Some v)] is [p v]. *)
+
+val exists : ('a -> bool) -> 'a option -> bool
+(** [exists p None] is [false],
+    [exists p (Some v)] is [p v]. *)
```
The `@since` values for those two new functions are missing.

gasche replied:
Thanks! (I split this change in #13920 for better discussion.)

gasche committed:
Changes

lthls commented:
```ocaml
module type T = sig type t val mk : unit -> t end
module F (X : T) = struct type t = { x : X.t } [@@unboxed] let mk () = { x = X.mk () } end
module F10 (X : T) = F(F(F(F(F(F(F(F(F(F(X))))))))))
module F100 (X : T) = F10(F10(F10(F10(F10(F10(F10(F10(F10(F10(X))))))))))

module B = struct type t = float let mk () = 0. end
module M = F(F100(B))

let x = M.mk ()
let y = lazy x
let a = Array.make 10 y
let z = Lazy.force a.(0)
```
Oops... (this is not actually caused by this PR, but I can extend it to a few more cases with this PR).

gasche commented:
I believe that you are pointing at a bug in the code that computes the "representation type" of values to decide whether short-cutting should be permitted. (This looks like an issue in the out-of-fuel behavior. Did you report it as a bug somewhere already?) This seems entirely orthogonal to the proposed change, isn't it? The PR uses the representation type to decide whether a structured value needs a forward block, just like trunk uses the representation type to decide if an identifier needs a forward block.

lthls commented:
I found the bug because I thought your PR introduced a bug around that, but I turned out to be wrong (because the type we use to decide whether we need a forward block or not does look through unboxed types), but I found this bug instead, which is entirely orthogonal to your PR indeed, and caused by a bad handling of the case where we run out of fuel in `Typedecl_unboxed.get_unboxed_type_representation`. I should probably submit an issue with that, indeed.

On another note, I found it quite surprising that you would keep the `` `Constant_or_function `` constructor name, given that it might be neither constant nor a function now.

gasche commented:
(Hehe. I also did some experiments around this feature, I would guess we both tried to see what happens to `type t = T of t Lazy.t ;; let rec th = lazy (T th)`.)

Yes, we could rename the constructor. At first I just reused what was already there, including `Ident Forward` for things that are not identifiers. Maybe we could just use `Value_can_shortcut | Value_needs_forward | Expr_thunk` ? (It's not clear that we get anything by special-casing floats.)

gasche commented:
Also: I thought of detecting structured values at the lambda level where this would be more convenient. This would work well for the usage of this classify function in Translcore, but not for the preliminary usages in Value_rec_check, where we don't want to compile the code just yet. (It's hard to play it safe in Value_rec_check because users assume that `let rec x = lazy e` is broadly supported, and we in theory could break code by short-cutting / thunking less. Over-approximating shortcut-ability can reject user code, and under-approximating it could accept unsafe recursive definitions.)

lthls reviewed (changes requested):
I'm broadly in favour of the feature proposed by this PR. However, there are a few things I want to see changed.
The first one is terminology. This PR mentions "constants" in various places, for things that I don't consider constants at all. The terms I'm used to for such expressions are "simple expressions" (what is a simple expression usually depends on context, so using this term here is not likely to be confused with other kinds of simple expressions such as the ones used in the middle-end). I would be also fine with "pure expressions", even though this might be a bit misleading (we don't want to lift all pure expressions out of lazy blocks). For this reason, the `` `Constant_or_function `` constructor needs to be renamed. I'm fine with @gasche's suggestion of `Value_can_shortcut | Value_needs_forward | Needs_thunk`, but I would slightly prefer something without "thunk" (maybe `General_case` or `Any_expression` for the last case).

The second thing I would like to see is an argument on why this doesn't cause some recursive definitions to become rejected. My own experiments make me relatively confident that we're not going to break anything serious, and I suspect that we're not going to break anything at all, but I'd like to see a good argument for that. Since @gasche has been heavily involved in `Value_rec_check`, I suspect he can produce such an argument (maybe even a formal version ?).

And as a last point, I think we should not consider as ~~constants~~ values expressions which are too large. With generated code in particular, I could see cases where users are building huge structures directly that may or may not be actually used, and choose to put the allocation under a `lazy` to prevent it from actually consuming memory (or CPU) before it is used. Users could fix this by adding `Sys.opaque_identity` or something similar, but I don't think they should have to do that.

As an extra comment (not something I expect to be taken into account for this PR), I think it would be better in the end to delay this optimisation of lazy evaluation to the middle-end. This would have two main advantages: the `Value_rec_check` wouldn't need to be aware of it (instead, the middle-end would ensure that lazy blocks that flow into `caml_update_dummy` are not shortcut), and it would apply after optimisations including inlining, so `lazy (List.map succ [0])` would become eligible too (the actually interesting cases are the ones mentioned in the description of this PR, but where `Cons` is only exported as a `cons` function).
This does lead into an issue though: how do we ensure in the middle-end that we don't shortcut lazy values that are expected not to be ?
It turns out that this is already an issue at the moment. Consider the following program:
```ocaml
let f x =
  let rec l =
    let v = lazy x in
    Gc.minor ();
    v
  in
  l
```
If we call `f 1`, we will get a segfault as it tries to update a dummy block of size 1 using the integer 1 (tagged).
I will submit an issue with that (and the other issue with fuel).

gasche commented:
Thanks for the excellent comments.

> This PR mentions "constants" in various places, for things that I don't consider constants at all.

Right, my idea at the start was to extend the previously accepted grammar (variables and shallow constant literals) with the application of variant constructors, and that's it. So when I started I really had "constants" in mind -- modulo inline records with mutable fields which I had not considered.

In fact, thinking about it, it's quite clear that we don't literally need a constant, we are trying to characterize expressions whose evaluation can safely be moved earlier in program execution (including if they would never occur otherwise). So they must (in flambda parlance) have no coeffects, and generative effects only -- generative effects are typically used to allow erasing some expressions, here we do the reverse transformation.

We could say that such an expression is "commutative", its evaluation can be reordered with the evaluation of any other expression in the same syntactic scope.

> The second thing I would like to see is an argument on why this doesn't cause some recursive definitions to become rejected.

In the general case `lazy e` makes recursive definitions happy because the evaluation of `e` is delayed (just like `fun _ -> e`), it is typed in Delay mode by Value_rec_check. In the particular case where the compiler decides to evaluate `e` right now, we use a less permissive mode, `Return`, the mode of sub-expressions whose value may be returned as the result of the recursive definition. (We could use `Guarded`, which is used for values placed a constructor, if we were confident that a forward block will be constructed, but I made the choice to be pessimistic about this and always return).

Moving a recursive variable from Delay mode to Return mode can break recursive definitions: `let rec x = lazy x` would be accepted in Delay mode, but is rejected in Return mode. This is already the case in trunk and older OCaml versions, the PR does not change this.

The PR does change cases of the form `lazy v`, where `v` is a structured constants that may contain variables (including recursive variables) placed under constructors, for example `let rec x = lazy Cons(1, x)`. In this case, the `lazy` context is downgraded from Delay to Return, but the recursive variable is then placed under a Guard context by the `Cons(1, _)` application, and so the whole expression becomes accepted again. The same phenomenon happens in general with all application of constructors (including tuples, records, etc.).

There are two categories of niche examples that build two categories of cases that build commutative terms with recursive variables that are not in Guarded position, but in Return or Dereference positions, and those could now be rejected under a `lazy`

```ocaml
(* example of Dereference usage of recursive variable [y] *)
type t = { field : t }
let rec x = lazy (Some y.field) and y = { field = y }

(* example of Return usage of recursive variable [x]; uses -rectypes *)
let rec x = lazy (let y = x in y)
```

Both examples are accepted in past OCaml versions and in trunk, and they would be rejected if the PR recognized them as commutative and optimized the Delay away. But it so happens that they both correspond to situations that are not handled by the analysis I implemented -- immutable field projections and let-bindings. So the PR does *not* reject them.

I think that those are niche examples and it would be okay to reject them. Users could fix their code by using `lazy (Sys.opaque_identity ...)` as you suggest, in the unlikely event that they decided to write something like this in the first place.

> And as a last point, I think we should not optimize commutative expressions which are too large.

Meh, what cutoff criterion would we implement? (I guess I could just count the number of recursive calls to `is_commutative` and cutoff at 42 or whatever.)

lthls commented:
> Meh, what cutoff criterion would we implement? (I guess I could just count the number of recursive calls to `is_commutative` and cutoff at 42 or whatever.)

I think that what you propose is fine, as far as I'm concerned. If I had to design a more logical criterion, I would estimate the size of both allocations (lazy block + closure vs sum of the various constructors involved) and use a formula like `init_alloc + use_probability * (lazy_force_cost + alloc_cost) > alloc_cost`, where `use_probability` would be around `0.5` and `lazy_force_cost` would be a constant of maybe 10 or 20 words (corresponding not to actual allocated words but to the time spent forcing a non-evaluated lazy value, divided by the average time spent for allocating one word). These numbers would probably have to be replaced by values obtained from actual benchmarks, but the idea is there. It would mean that allocations of the form `(x1, x2, ...., xn)` are likely eligible for forwarding, as allocating the closure is about as expensive as allocating the value itself, but `(x, x, ..., x)` or `(x, const, ..., const)` are likely to be compiled to a small closure allocation instead of a large structure.

gasche commented:
I defined a simpler variant type `lazy_summary`, and renamed `is_constant_expr` into `small_and_commutative`. I think that this addresses the main comments that are in scope for this PR. (I'm thinking about `caml_update_dummy_lazy`, as a separate change.)

gasche commented:
@lthls I rebased this PR on top of trunk (and squashed the intermediary commits together). I think that we should ask ourselves again if we want static optimizations for the lazy-constructor case (the case where lazy data structures are built strictly; see microbenchmarks for demonstrated performance benefits). If not, we can also close the PR.

(I thought of this PR while studying an experimental programming language's implementation of lazy values, and noticing that the reason why OCaml did worse in a particular situation was due to this missing optimization.)

lthls reviewed (approved):
I don't feel strongly about the need for the optimisation, but I have read the PR again and I think it is worth merging for the nice cleanup and documentation.

I think that the way you compute size is a bit off (in the record case, you only consider the overridden fields, so huge record updates where only one field is changed count as a single unit of size), but it's not a blocking issue for me.

gasche committed:
fixup! (lazy e): optimize structured constants as Lazy.from_val would do

gasche commented:
Thanks! (I updated the size-computation for `Kept` record fields.)
